Make {integer}::from_str_radix constant

GKFX opened the pull request:
This commit makes FromStr on integers constant so that `const x: u32 = "23".parse();` works. More practical use-case is with environment variables at build time as discussed in https://github.com/rust-lang/rfcs/issues/1907.

Tracking issue #59133.

ACP: https://github.com/rust-lang/libs-team/issues/74

rust-highfive commented:
r? @thomcc

(rust-highfive has picked a reviewer for you, use r? to override)

GKFX commented:
@rustbot label +T-libs-api -T-libs

GKFX commented on library/core/src/cmp.rs:
```diff
@@ -1384,7 +1384,8 @@ mod impls {
             #[stable(feature = "rust1", since = "1.0.0")]
-            impl PartialOrd for $t {
+            #[rustc_const_unstable(feature = "const_primitive_ords", issue = "none")]
+            impl const PartialOrd for $t {
```
All changes in this file redundant to #92390.

thomcc replied:
Let's wait for that to land first too, then.

GKFX commented on library/core/src/fmt/mod.rs:
```diff
@@ -315,7 +315,7 @@ macro_rules! arg_new {
         #[unstable(feature = "fmt_internals", reason = "internal to format_args!", issue = "none")]
         #[inline]
-        pub fn $f<'b, T: $t>(x: &'b T) -> ArgumentV1<'_> {
+        pub const fn $f<'b, T: $t>(x: &'b T) -> ArgumentV1<'_> {
```
Unless there's an existing pull request I haven't spotted, I'm expecting to need to spin this and src/test/ui/consts/const-eval/format.* off to a second pull request.

GKFX replied:
This doesn't actually work (the const panic formatting code breaks) so I had to do the formatting by hand using `fmt_u64` from `fmt/num.rs`.

thomcc commented:
Can you file an API change proposal, as requested by the bot? Thanks.

> If this PR contains changes to any unstable APIs please edit the PR description to add a link to the relevant [API Change Proposal](https://std-dev-guide.rust-lang.org/feature-lifecycle/api-change-proposals.html) or [create one](https://github.com/rust-lang/libs-team/issues/new?assignees=&labels=api-change-proposal%2C+T-libs-api&template=api-change-proposal.md&title=%28My+API+Change+Proposal%29) if you haven't already

GKFX commented:
Done: https://github.com/rust-lang/libs-team/issues/74

thomcc reviewed (commented):
~~The implementation looks largely fine, so r=me pending api review.~~ The implementation has changed enough that I'd like to re-review.

bors commented:
:umbrella: The latest upstream changes (presumably #99451) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

klensy commented on library/core/src/num/mod.rs:
```diff
@@ -1074,13 +1089,15 @@ fn from_str_radix<T: FromStrRadixHelper>(src: &str, radix: u32) -> Result<T, Par
-                for &c in digits {
+            ($unchecked_additive_op:expr) => {{
+                let mut i = 0; // C-style iteration due to lack of const_for.
+                while let Some(&c) = digits.get(i) {
```
Is this adds\removes bound checks, and generally, perf changes?

This ( and lower one) can be rewritten without `i` like `while let Some((first, last)) = digits.split_first()`

GKFX replied:
Looks like one bounds check per iteration to me (in `get()` only) and [Godbolt suggests](https://godbolt.org/#z:OYLghAFBqd5TKALEBjA9gEwKYFFMCWALugE4A0BIEAZgQDbYB2AhgLbYgDkAjF%2BTXRMiAZVQtGIHgBYBQogFUAztgAKAD24AGfgCsp5eiyahSAVyVFyKxqiIEh1ZpgDC6embZMpATnLOAGQImbAA5TwAjbFIQADZyAAd0JWIHJjcPL19E5NShIJDwtiiY%2BJtsOzSRIhZSIgzPbx4/csqhatqiArDI6LjrGrqGrOaBzu6ikriASmt0M1JUTi5sdSJopgBqAFIAdgAhba0AQU2zzZotwXQITU2zAA5p7YBmQ5O9gBEj45%2BEswiFy26ggERAOwATLFtgBWfaPWGfaY7A4/c6bRhETZsMxYgg7F6fTZaV7vU7nADuSAY2Ax2CxInQHAg2yhEQAnutka8iREAHTAekQAjc1EndHosxMJQsGi0vb7C7oG4crko75vNESs741mHQmbHikrVnL4/L5cWb0bgw/jeLg6cjobgAJQsWKU80W8ohLz45CI2ktswA1iAXhC%2BRCHloeA9YjwXrtYtIXg8fDDDNxpPw2OHdnzpLs49JYg9S%2BWy5n7Y7nVx%2BEoQFoA0HZnBYCgMDh8MQyJRqHRGKwONx/YJhGIJJwZHJhMo1JoHXoQBDDMZTO7rNhbPZHBBnMNvKvAsEesU%2Bn4kild%2Bl3I0Vzlr2kJr0Yi1txUbx0hneRlud1UgxdKekx9DwYw/pkTQQcBhSvlIsyegsSzcKs6ykFsComkCSo3HcjzPJqHy7N8HwnP8gKXJsIJgpC0JwgiMJIiiZLooIpB0aqtLBJsgJYeK2r3NKsrygcuGgpy2CihqrHnGaxGfJaWZcLa5A1vwdZupYmxId6kIRvwgZLtMVrZrm4ZaHyMKxDCLwwjC8YPC8zm7FosjqU63ANk2LbGeQ7bIGgWB4IQJAUFQtA0sOyxjvIk6SDO46KCoGhBuQ%2BgvGuJggOYlj/p%2BaROEwri/ke/jFS%2B54xDwzZXnkt5QQYdU3pVUyZq0X5AYeD4dYB4wgfBNUwd14GWP1cFVSAMKIV6KErGsGwsdhVHXLc4IEcaCnmuRAI4SCOJYrRrL0fCDyIqKslnFSNJ0lisL7Fx5CbKQ2DaQAApsfJ8oiBK8ktAnalKMpyix4kAFRcdJm3ktqfEGi9ljQ%2Bi8m/CRSnWipdppZp7o6bNPrOYZrazEg2AsDgMQQKZXA5uQeZ%2BnyPixBCuw%2BLs0bSA8zTSNI1bY151g%2BUZOgmeQYbSD4Vl%2Bj4Ly86zsQyw5ykvFjS6efWvki/5iABRAnZMgkNLhdQGBsIbjAxKQPDM82g7oY2oJpREwS1Oyo78KbHDCAA8kw9Bu2rOBsOukiBwQL2VAAbq9aWrBUuIxfwwTrBjjr0AQESkK7bg4O7AakAQea8EpNBGMASgAGoENgFLewkzB50l8XTrISXzqlav6OBRjZblVjpxEjaQLM6AJDejZcAAtC4myj0Qk%2BMNH9A8ir%2BVtN4%2B7FSN5WYK1YGPvV2/Nc%2BA2TaNH7r9%2B9SlQYvXtEBe/VcNN%2BjQ/p9TDwM3IdO%2BfYIn1OqQ8nWPuqBDR8mZnyLQmwIBaSsJsHsYV9LgU2G4M2RskHTCJsZEmZMKbUFDFNSyrkIRaB8FzF4bkYRM0yhjWm9MCx8zVnWbyzZhZKV1vrNBFt%2BwQFNubPoVsbYCAYPbagYI1bO1YKQAO/pPbMCIL7f2aUg4h2WI6QgEd7DRwno6OOqAE552TtuNKA8s7SJzmowyBci58FmKXFg5cq41zrg3Yus5RDiASq3eQ7dFyOn0KuHuG48oDyHlTJ0Y80gT2nrPMeC9sBLxXg2C%2BN4iolUauBE8E0P4HxvEfXILV377zvkwK%2B28SlX0frfLqL8YJVM/nMb%2BD4iAvX/spQB/MuAgLARAqBMDNzwNCmQfSq4UEG3QayCMmDNbBnIKTcmfRwm0PMhQvkMgfBaBhOszZ2zdhqU6SwmZbYdacP4cbXh4yLY5WtiQ4R9BRGOwkS7aRec5E%2Bz9gHdR2Bg4mFDuo8OBVtGx3UPHdYhjhDGLVqY7OwU0otMLu7OxZdK7V1rvXe0sUJyeJbu43xaUMpZWCf3DOYSR6RKEBPZJAE9wHlqVks8OTj5CHyU%2BfIRSn4VJqRktenVxoMuKVy%2B8r8%2BWgWql/b04Zf5tIxh0ph3Buk8HAZGPpsCnoIOGZMzKYyuHRH0i8aZbDRbzLwUssydNwyRjsqQzmMJrUPC2fsuVGtGysOJmLQhhYyxaBstIEhFCUwQlkBjVeQCBaGuphCVWtYw1uujqQFIjhpBAA) that the version with `i` compiles to the same as `for &c in digits`, but the pattern matching one doesn't get optimized so well.

klensy replied:
In your link you missed `-C opt-level=3` for 3rd code example.

GKFX replied:
Thanks - they are all the same then. Personally I think the version with `i` is less complicated to read but no firm preference.

GKFX commented on library/core/src/fmt/num.rs:
```diff
@@ -478,93 +486,6 @@ mod imp {
 impl_Exp!(i128, u128 as u128 via to_u128 named exp_u128);
 
-/// Helper function for writing a u64 into `buf` going from last to first, with `curr`.
-fn parse_u64_into<const N: usize>(mut n: u64, buf: &mut [MaybeUninit<u8>; N], curr: &mut isize) {
```
This function was effectively a duplicate of fmt_u64 and seemed slower in a quick benchmark I ran, so I de-duplicated and replaced it with `fmt_u64`.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/7474412819?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/5f73b022ad9a3fdb29f96563348331af886229b4/checks/7474412819/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

---- [ui] src/test/ui/consts/const-eval/parse_ints.rs stdout ----
diff of stderr:

5 LL | |     "0x1234567890123456", "0x5634129078563412", "0x6a2c48091e6a2c48",
6 LL | |     "[0x56, 0x34, 0x12, 0x90, 0x78, 0x56, 0x34, 0x12]",
7 LL | |     "[0x12, 0x34, 0x56, 0x78, 0x90, 0x12, 0x34, 0x56]",
- LL | |     "", ""}
-    | |___________- inside `core::num::<impl u64>::from_str_radix` at $SRC_DIR/core/src/num/uint_macros.rs:LL:COL
+ LL | |     "", "", ""}
+    | |_______________- inside `core::num::<impl u64>::from_str_radix` at $SRC_DIR/core/src/num/uint_macros.rs:LL:COL
11 LL |           panic!("{}", {
12    |  _________^


28 LL |   const _TOO_LOW: () = { u64::from_str_radix("12345ABCD", 1); };
29    |                          ----------------------------------- inside `_TOO_LOW` at $DIR/parse_ints.rs:7:24
-    = note: this error originates in the macro `$crate::panic::panic_2021` (in Nightly builds, run with -Z macro-backtrace for more info)
+    = note: this error originates in the macro `$crate::panic::panic_2021` which comes from the expansion of the macro `panic` (in Nightly builds, run with -Z macro-backtrace for more info)
32 
33 error[E0080]: evaluation of constant value failed
33 error[E0080]: evaluation of constant value failed
34   --> $SRC_DIR/core/src/num/mod.rs:LL:COL

37 LL | |     "0x1234567890123456", "0x5634129078563412", "0x6a2c48091e6a2c48",
38 LL | |     "[0x56, 0x34, 0x12, 0x90, 0x78, 0x56, 0x34, 0x12]",
39 LL | |     "[0x12, 0x34, 0x56, 0x78, 0x90, 0x12, 0x34, 0x56]",
- LL | |     "", ""}
-    | |___________- inside `core::num::<impl u64>::from_str_radix` at $SRC_DIR/core/src/num/uint_macros.rs:LL:COL
+ LL | |     "", "", ""}
+    | |_______________- inside `core::num::<impl u64>::from_str_radix` at $SRC_DIR/core/src/num/uint_macros.rs:LL:COL
43 LL |           panic!("{}", {
44    |  _________^


60 LL |   const _TOO_HIGH: () = { u64::from_str_radix("12345ABCD", 37); };
61    |                           ------------------------------------ inside `_TOO_HIGH` at $DIR/parse_ints.rs:8:25
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
-    = note: this error originates in the macro `$crate::panic::panic_2021` (in Nightly builds, run with -Z macro-backtrace for more info)
+    = note: this error originates in the macro `$crate::panic::panic_2021` which comes from the expansion of the macro `panic` (in Nightly builds, run with -Z macro-backtrace for more info)
64 
---
To only update this specific test, also pass `--test-args consts/const-eval/parse_ints.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/consts/const-eval/parse_ints.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/consts/const-eval/parse_ints" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/consts/const-eval/parse_ints/auxiliary"
stdout: none
--- stderr -------------------------------
  --> /checkout/library/core/src/num/mod.rs:1055:9
   |
   |
LL | /     uint_impl! { u64, u64, i64, NonZeroU64, 64, 18446744073709551615, 12, "0xaa00000000006e1", "0x6e10aa",
LL | |     "0x1234567890123456", "0x5634129078563412", "0x6a2c48091e6a2c48",
LL | |     "[0x56, 0x34, 0x12, 0x90, 0x78, 0x56, 0x34, 0x12]",
LL | |     "[0x12, 0x34, 0x56, 0x78, 0x90, 0x12, 0x34, 0x56]",
LL | |     "", "", ""}
   | |_______________- inside `core::num::<impl u64>::from_str_radix` at /checkout/library/core/src/num/uint_macros.rs:68:13
LL |           panic!("{}", {
   |  _________^
   | |_________|
   | |
   | |
LL | |             let buf_end = buf.as_mut_ptr_range().end;
LL | |             let radix_len = crate::fmt::fmt_u64(
LL | |                 radix.into(),
LL | |             }
LL | |         });
   | |          ^
   | |__________|
   | |__________|
   | |__________the evaluated program panicked at 'from_str_radix_int: must lie in the range `[2, 36]` - found 1', /checkout/library/core/src/num/mod.rs:1055:9
   |            inside `core::num::from_str_radix::<u64>` at /checkout/library/core/src/panic.rs:54:9
  ::: /checkout/src/test/ui/consts/const-eval/parse_ints.rs:7:24
   |
   |
LL |   const _TOO_LOW: () = { u64::from_str_radix("12345ABCD", 1); };
   |                          ----------------------------------- inside `_TOO_LOW` at /checkout/src/test/ui/consts/const-eval/parse_ints.rs:7:24
   = note: this error originates in the macro `$crate::panic::panic_2021` which comes from the expansion of the macro `panic` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0080]: evaluation of constant value failed
  --> /checkout/library/core/src/num/mod.rs:1055:9
  --> /checkout/library/core/src/num/mod.rs:1055:9
   |
LL | /     uint_impl! { u64, u64, i64, NonZeroU64, 64, 18446744073709551615, 12, "0xaa00000000006e1", "0x6e10aa",
LL | |     "0x1234567890123456", "0x5634129078563412", "0x6a2c48091e6a2c48",
LL | |     "[0x56, 0x34, 0x12, 0x90, 0x78, 0x56, 0x34, 0x12]",
LL | |     "[0x12, 0x34, 0x56, 0x78, 0x90, 0x12, 0x34, 0x56]",
LL | |     "", "", ""}
   | |_______________- inside `core::num::<impl u64>::from_str_radix` at /checkout/library/core/src/num/uint_macros.rs:68:13
LL |           panic!("{}", {
   |  _________^
   | |_________|
   | |
   | |
LL | |             let buf_end = buf.as_mut_ptr_range().end;
LL | |             let radix_len = crate::fmt::fmt_u64(
LL | |                 radix.into(),
LL | |             }
LL | |         });
   | |          ^
   | |__________|
   | |__________|
   | |__________the evaluated program panicked at 'from_str_radix_int: must lie in the range `[2, 36]` - found 37', /checkout/library/core/src/num/mod.rs:1055:9
   |            inside `core::num::from_str_radix::<u64>` at /checkout/library/core/src/panic.rs:54:9
  ::: /checkout/src/test/ui/consts/const-eval/parse_ints.rs:8:25
   |
   |
LL |   const _TOO_HIGH: () = { u64::from_str_radix("12345ABCD", 37); };
   |                           ------------------------------------ inside `_TOO_HIGH` at /checkout/src/test/ui/consts/const-eval/parse_ints.rs:8:25
   = note: this error originates in the macro `$crate::panic::panic_2021` which comes from the expansion of the macro `panic` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 2 previous errors

```

</details>

fee1-dead commented on library/core/src/num/mod.rs:
```diff
@@ -1033,19 +1034,41 @@ impl_helper_for! { i8 i16 i32 i64 i128 isize u8 u16 u32 u64 u128 usize }
+    if 2 > radix || radix > 36 {
+        // FIXME DIY const string formatting. Doing it the normal way is blocked on something
+        // like https://github.com/rust-lang/rust/issues/86364.
+        const MESSAGE: &[u8; 60] = b"from_str_radix_int: must lie in the range `[2, 36]` - found ";
+        let mut buf = [mem::MaybeUninit::<u8>::uninit(); MESSAGE.len() + 10];
+        panic!("{}", {
+            let buf_end = buf.as_mut_ptr_range().end;
+            let radix_len = crate::fmt::fmt_u64(
+                radix.into(),
+                // SAFETY: OK to refer to the last 20 bytes of a 70 byte buffer.
+                unsafe { &mut *(buf_end.sub(20).cast()) },
+            )
+            .len();
+            let result = &mut buf[(10 - radix_len)..];
+            // SAFETY: (10-radix_len) above is in 1..=10 so 60 <= result.len() <= 69
+            unsafe {
+                MESSAGE.as_ptr().copy_to_nonoverlapping(result.as_mut_ptr().cast(), MESSAGE.len());
+                // SAFETY: result is entirely filled by MESSAGE and the radix, both ASCII.
+                crate::str::from_utf8_unchecked(mem::MaybeUninit::slice_assume_init_ref(result))
+            }
+        });
+    }
```
FYI, there is also the way of using const_eval_select: https://github.com/rust-lang/rust/blob/f80e454450c891c075617d23532c6eb317f4471d/library/core/src/slice/index.rs#L34-L57

bors commented:
:umbrella: The latest upstream changes (presumably #99735) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

GKFX commented:
@rustbot ready
Assuming the CI passes I think this is now in better shape. https://github.com/rust-lang/rust/pull/92390 got merged so I have removed that code, and the use of const_eval_select avoided the need to figure out how to do const-compatible string formatting so those changes are also gone.

thomcc commented:
I'm not sure if making something like this constified is doable yet, so CC @rust-lang/wg-const-eval (and see also https://github.com/rust-lang/libs-team/issues/74)

oli-obk commented on src/test/ui/consts/const-eval/parse_ints.rs:
```diff
@@ -0,0 +1,10 @@
+#![feature(const_int_from_str)]
+
+const _: () = match "-1234".parse::<i32>() {
+    Ok(x) => assert!(x == -1234),
+    Err(_) => panic!(),
+};
+const _TOO_LOW: () = { u64::from_str_radix("12345ABCD", 1); };
+const _TOO_HIGH: () = { u64::from_str_radix("12345ABCD", 37); };
```
you need to add `//~ ERROR message goes here` annotations to make the test pass

Please also add a `// check-pass` test that shows successful usages of this in a const item initializer.

RalfJung replied:
> Please also add a // check-pass test that shows successful usages of this in a const item initializer.

Or better, add these tests in the libcore test suite. ui tests should only be used when regular unit/integration tests are insufficient.

GKFX replied:
> you need to add //~ ERROR message goes here annotations to make the test pass

The test passes fine already; is there some situation in which it is required.

>successful usages of this in a const item initializer.

I have added a const item initializer, `_TEST_CONST_PARSE`, to the libcore test suite.

thomcc commented:
According to https://github.com/rust-lang/libs-team/issues/74#issuecomment-1225163528 we probably don't want to do this yet. Can this instead be change to instead just be applied to `from_str_radix`?

fee1-dead reviewed (approved):
I'm fine with the `const` trait impls for testing the experimental feature in the library. If you would like to see `from_str_radix` to be stabilized, it is better if it does not depend on `const_trait_impl`. But since wg-const-eval can approve PRs adding unstable const impls, I would expect only some minor changes to `from_str_radix` to be needed. (and preferably not removing the const impls)

JohnCSimon commented:
Ping from triage: 
@GKFX what is the status of this PR?

FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

GKFX commented:
I've removed the const implementations of FromStr as requested at the https://github.com/rust-lang/libs-team/issues/74#issuecomment-1225163528 so this now just does `{integer}::from_str_radix`, which is more versatile anyway.

faptc commented on library/core/src/num/mod.rs:
```diff
@@ -1105,19 +1106,45 @@ impl_helper_for! { i8 i16 i32 i64 i128 isize u8 u16 u32 u64 u128 usize }
+#[cold]
+#[track_caller]
+const fn from_str_radix_assert(radix: u32) {
+    if 2 > radix || radix > 36 {
```
Is this condition correct?

GKFX replied:
Yes; it is the inverse of the previous condition `(2..=36).contains(&radix)`.

faptc replied:
I think the inverse would be `2 < radix` ...

GKFX replied:
The original condition was `2 <= radix`. `!(2 <= radix)` equals `2 > radix`. `2 < radix` is neither of those so not correct.

faptc replied:
You're right, I am just used to the style `a < n` instead.

GKFX commented:
@rustbot ready

GKFX commented:
NB that `ParseIntError::kind` was made const to allow errors to be read from const item initializers, as I noticed that was otherwise impossible. There is also technically no longer a tracking issue for this PR as #59133 referred specifically to `FromStr`.

thomcc commented:
Can you rebase this? It basically looks fine to me.

@rustbot author

fee1-dead commented:
This isn't currently possible, as we removed the ability to have const impls for traits in the standard library.

@rustbot -S-waiting-on-author +S-blocked

thomcc commented:
@fee1-dead This is just about the https://doc.rust-lang.org/nightly/std/primitive.i64.html#method.from_str_radix functions, which are not tied to traits.

tarcieri commented:
@thomcc the implementation appears to internally make use of const impls of traits, however

thomcc commented:
Ah, well, it will need to change that.

thomcc commented:
If the author wants to update this to use macros instead of const traits (similar to how we avoid this kind of duplication in the rest of libcore), I think there's a version of that PR we'd take (if it's done reasonably).

GKFX committed:
Make {integer}::from_str_radix constant

GKFX commented:
I've updated this MR to remove internal usage of const traits and rebased.

In order to avoid creating a large number of different method names from_str_radix_u8, from_str_radix_u16, etc and making the macros that consume them more complicated and error-prone, I implemented {integer}::from_str_radix directly in num/mod.rs and removed them from `[u]int_macros.rs`.

Looping over the slice in const code without introducing duplicate bounds checks was also made more challenging by `[u8]::get` no longer being const, which is the reason for `const_slice_get`.

GKFX commented:
@rustbot ready
Let me know if this is OK.

GKFX commented:
@thomcc Are there any more changes needed on this or can it be merged?

bors commented:
:umbrella: The latest upstream changes (presumably #117482) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

thomcc commented:
I'm going to be away for a few months, so I'm rerolling my PRs so that folks don't have to wait for me. Sorry/thanks.

r? libs

m-ou-se commented:
r? libs

Mark-Simulacrum commented on library/core/src/num/error.rs:
```diff
@@ -113,8 +113,9 @@ pub enum IntErrorKind {
 impl ParseIntError {
     /// Outputs the detailed cause of parsing an integer failing.
     #[must_use]
+    #[rustc_const_unstable(feature = "const_int_from_str", issue = "none")]
```
Please update this with a tracking issue number (I think reusing the same one as for parse_str is fine).

Mark-Simulacrum commented on library/core/src/num/mod.rs:
````diff
@@ -1372,144 +1371,222 @@ pub enum FpCategory {
+        #[doc = concat!("assert_eq!(", stringify!($size), "::from_str_radix(\"A\", 16), Ok(10));")]
+        /// ```
+        #[stable(feature = "rust1", since = "1.0.0")]
+        #[rustc_const_unstable(feature = "const_int_from_str", issue = "none")]
````
Tracking issue here too.

Mark-Simulacrum reviewed (commented):
r=me with tracking issues populated and a rebase

GKFX commented:
@rustbot ready

Mark-Simulacrum commented on library/core/src/num/mod.rs:
```diff
@@ -1386,144 +1385,216 @@ pub enum FpCategory {
-        radix
-    );
+fn from_str_radix_panic_rt(radix: u32) -> ! {
+    panic!("from_str_radix_int: must lie in the range `[2, 36]` - found {}", radix);
```
Shouldn't both of these helpers have track_caller?

GKFX commented:
@rustbot ready

Mark-Simulacrum commented:
Needs a rebase, but r=me on current state.

GKFX commented:
@rustbot ready

Rebased again.

Mark-Simulacrum commented:
@bors r+

GKFX commented:
@Mark-Simulacrum Bors doesn't seem to have responded to your comment and this PR doesn't seem to be visible at https://bors.rust-lang.org/queue/rust. Is it stuck in some way?

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit 3855b8bb609bbaaa3871797b40a78757f88d9b12 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 3855b8bb609bbaaa3871797b40a78757f88d9b12 -->
<!-- homu: {"type":"Approved","sha":"3855b8bb609bbaaa3871797b40a78757f88d9b12","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 3855b8bb609bbaaa3871797b40a78757f88d9b12 with merge 8df7e723ea729a7f917501cc2d91d640b7021373...
<!-- homu: {"type":"BuildStarted","head_sha":"3855b8bb609bbaaa3871797b40a78757f88d9b12","merge_sha":"8df7e723ea729a7f917501cc2d91d640b7021373"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8493042299/job/23267897578)
Approved by: Mark-Simulacrum
Pushing 8df7e723ea729a7f917501cc2d91d640b7021373 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8493042299/job/23267897578"},"merge_sha":"8df7e723ea729a7f917501cc2d91d640b7021373"} -->

rust-timer commented:
Finished benchmarking commit (8df7e723ea729a7f917501cc2d91d640b7021373): [comparison URL](https://perf.rust-lang.org/compare.html?start=70714e38f224ef1d50f3f772808fff65d7a29c0b&end=8df7e723ea729a7f917501cc2d91d640b7021373&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=70714e38f224ef1d50f3f772808fff65d7a29c0b&end=8df7e723ea729a7f917501cc2d91d640b7021373&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=70714e38f224ef1d50f3f772808fff65d7a29c0b&end=8df7e723ea729a7f917501cc2d91d640b7021373&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=70714e38f224ef1d50f3f772808fff65d7a29c0b&end=8df7e723ea729a7f917501cc2d91d640b7021373&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.2% | [2.2%, 2.2%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=70714e38f224ef1d50f3f772808fff65d7a29c0b&end=8df7e723ea729a7f917501cc2d91d640b7021373&stat=size:linked_artifact)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.0%  | [0.0%, 0.0%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.4% | [-0.8%, -0.0%] | 12    |
| Improvements ✅ <br /> (secondary) | -0.4% | [-1.3%, -0.0%] | 72    |
| All ❌✅ (primary)                 | -0.4% | [-0.8%, -0.0%] | 12    |
</details>

**Bootstrap**: 667.994s -> 669.242s (0.19%)
**Artifact size**: 315.79 MiB -> 315.71 MiB (-0.03%)
